tests/functional/test_credentials.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from tests import unittest, mock, IntegerRefresher
+import datetime
+import threading
+import math
+import time
+
+from dateutil.tz import tzlocal
+
+from botocore import credentials
+from botocore import utils
+
+
+class TestCredentialRefreshRaces(unittest.TestCase):
+    def assert_consistent_credentials_seen(self, creds, func):
+        collected = []
+        threads = []
+        for _ in range(20):
+            threads.append(threading.Thread(target=func, args=(collected,)))
+        start = time.time()
+        for thread in threads:
+            thread.start()
+        for thread in threads:
+            thread.join()
+        for triple in collected:
+            self.assertTrue(triple[0] == triple[1] == triple[2], triple)
```

Comment:
Might be good to explain what is in the triple and why you are checking that they are all equal. It is because we keep track of how many times refresh is called and the number of refresh times all need to be the same in order to ensure that the credentials set all came from the same refresh. Right?